Updates to support offline payment methods.

dmallory42 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #56364  

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

**Before**
<img width="1329" alt="Screenshot 2025-04-29 at 17 11 10" src="https://github.com/user-attachments/assets/cd7f9297-8c78-4455-b9db-2f68bee84256" />

<img width="1346" alt="Screenshot 2025-04-29 at 17 11 20" src="https://github.com/user-attachments/assets/0636ae1d-18b2-4b60-8990-aae1eeee2e79" />

<img width="1344" alt="Screenshot 2025-04-29 at 17 11 27" src="https://github.com/user-attachments/assets/3195d19c-f05c-4db4-a0c1-381861fbbf59" />

**After**
<img width="1275" alt="Screenshot 2025-04-29 at 18 21 53" src="https://github.com/user-attachments/assets/bad93bd4-78d3-48e3-95b7-74ba5a429382" />

<img width="1288" alt="Screenshot 2025-04-29 at 18 22 33" src="https://github.com/user-attachments/assets/75e708a2-15d0-4108-a286-3f45c2eda123" />

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Spin up new JN site with this branch.
2. Ensure the new payment settings experience is enabled.
3. Go to `Settings > Payments > Offline Methods`.
4. Go to the `Direct bank transfer` method, verify the form loads. Make some changes to the form and save. Verify the form reloads and a successful `POST` request is sent. Reload the page and verify that your changes persisted.
5. Disable the new payment settings experience and go to the _old_ version of the direct bank transfer page. Verify that the changes you made using the new form are correctly persisted.
6. Repeat steps 4-5 with the other offline methods, cash on delivery and check. 

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

dmallory42 committed:
Updates to support offline payment methods.

dmallory42 committed:
Offline methods updates.

dmallory42 committed:
Merge remote-tracking branch 'origin/trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Lots of updates and fixes to the offline methods pages.

dmallory42 committed:
Lots of updates and fixes to the offline methods pages.

dmallory42 committed:
Lots of updates and fixes to the offline methods pages.

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Reformat.

dmallory42 committed:
Formatting.

dmallory42 committed:
Removing logging.

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Update to the types

dmallory42 committed:
Add changelog.

dmallory42 committed:
Add changelog

dmallory42 committed:
Fix indentation in SCSS file

dmallory42 committed:
Fix indentation in SCSS file

dmallory42 committed:
Fix indentation in SCSS file

dmallory42 committed:
Fix indentation in SCSS file

dmallory42 committed:
Fix indentation in SCSS file

dmallory42 committed:
Lint fix.

dmallory42 commented on plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-payment-gateways-controller.php:
```diff
@@ -223,4 +223,52 @@ public function get_item_schema() {
+	 * @param array        $setting The field settings.
+	 * @return array|WP_Error
+	 */
+	public function validate_setting_multiselect_field( $values, $setting ) {
```
This fixes an issue where the validation of options keys wasn't working as expected (because the array was multidirectional, but it was expecting a flat array)

dmallory42 committed:
Add needed scss class.

dmallory42 committed:
Fix title.

oaratovskyi commented:
**Testing-only review** (code review will follow soon)

Issues found:

- Cash on delivery can't be opened
![image](https://github.com/user-attachments/assets/f5894ec1-6b1d-4598-91cf-613257ee3dfd)
- Different columns in bank account details
![image](https://github.com/user-attachments/assets/a8f6e5df-1229-410f-abcb-4d77ab1acc18)
- Can't add bank account. Filled in the data but it wasn't saved. Tried to load a gif it's too big, will add to Slack.
![image](https://github.com/user-attachments/assets/9d39abc1-25ff-4aa6-8af6-4e603b4a9644)
  - do we need the country? no country in old version of the page
  - view/edit, delete ellipsis behavior and styling isn't consistent with what we have in NOX settings page (should hide when clicked)
- When editing bank account details saved from the old page, it's impossible to save
![image](https://github.com/user-attachments/assets/41262b11-d327-46cb-922c-2837211c3805)
  - IBAN field wasn't required in old version, now it is
  - required fields errors are not visible to user, they are not red

And this separately:
- Need to check out with the design, but I guess we need it consistent with other pages
![image](https://github.com/user-attachments/assets/b9264f07-0bf1-4dea-8ce7-2cb1d81b6226)
  - navigation bar should probably be hidden (or not - since it's not hidden in WooPayments settings)
  - top level text should be settings (it's everywhere like that) and the back button should be below (like in Fraud page)
  ![image](https://github.com/user-attachments/assets/48485f86-f2a8-4ebd-85a9-f2c07caeff47)

oaratovskyi commented on packages/js/data/src/payment-gateways/types.ts:
```diff
@@ -7,26 +7,32 @@ export type SettingDefinition = {
-	settings_url: string;
-	needs_setup?: boolean;
-};
+export type PaymentGateway< TSettings = Record< string, SettingDefinition > > =
```
It should work fine, but I wonder if this can break some existing functionality that relies on this type? Could you double check please? 🙏

dmallory42 replied:
The previous definition is incorrect 🙂 so this is really a bug fix to match the actual structure which gets returned. It was defining it like this (example):

```
settings: {
  default: string;
  description: string;
  id: string;
  label: string;
  options?: Record< string, Record< string, string > >;
  placeholder: string;
  tip: string;
  type: string;
  value: string;
  value: string | string[];
  is_dismissed: string;
}
``` 

when it should be more like:

```
settings: {
  "setting_1": {
    default: string;
    description: string;
    id: string;
    label: string;
    options?: Record< string, Record< string, string > >;
    placeholder: string;
    tip: string;
    type: string;
    value: string;
    value: string | string[];
    is_dismissed: string;
 },
 "setting_2": {
    default: string;
    description: string;
    id: string;
    label: string;
    options?: Record< string, Record< string, string > >;
    placeholder: string;
    tip: string;
    type: string;
    value: string;
    value: string | string[];
    is_dismissed: string;
 },
}
```

I will double check it didn't break anything else though, thanks for mentioning it

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/bank-accounts-list/bank-account-modal.tsx:
```diff
@@ -41,22 +46,18 @@ export const BankAccountModal = ( {
 	onSave,
 	defaultCountry,
 }: Props ) => {
+	const countries = window.wcSettings.countries;
```
Do we need those countries? I see that there is no countries in the old version. As far as I get, countries influence what fields are required, right? Could you link me please to the requirement from product/design about it, I seem to have missed it

dmallory42 replied:
Yes, country selection is a requirement from the designs by Magda (I didn't add it in the previous PR because it was just easier to address it in this one) - here's the P2 post which outlines the requirements: pfHfG4-17j-p2 🙂

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/bank-accounts-list/types.ts:
```diff
@@ -1,9 +1,7 @@
 export interface BankAccount {
-	id: string;
```
not sure what is the cause, but bank account creation does not work. Maybe due to this id absense?

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/bank-accounts-list/utils.ts:
```diff
@@ -1,10 +1,32 @@
-	if ( country === 'AU' ) return 'sort_code';
-	return 'iban';
+export const getSortcodeLabel = ( country: string ) => {
+	if ( country === 'AU' ) return __( 'BSB', 'woocommerce' );
```
I suggest to change this to `switch` 🙂

dmallory42 replied:
Good suggestion. Updated in c91f889324!

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-layout/payment-settings-layout.scss:
```diff
@@ -0,0 +1,22 @@
+	display: flex;
+	flex-flow: column;
+
+	// TODO: Add mobile styling (row instead of column).
```
spotted that, adding the comment, so it's not missed 🙈

dmallory42 replied:
Handled in 370a44a843!

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-layout/payment-settings-layout.scss:
```diff
@@ -0,0 +1,22 @@
+.payment-settings-layout {
+	margin: 24px auto 0;
+	max-width: 1032px;
```
is that so in design? number seems unusual

dmallory42 replied:
Have removed this constraint in 370a44a843. Good spot!

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-layout/payment-settings-layout.tsx:
```diff
@@ -0,0 +1,14 @@
+ */
+import './payment-settings-layout.scss';
+
+export const PaymentSettingsLayout = ( {
```
could you add some small comment about this component? It's not self-explanatory for me

oaratovskyi replied:
maybe PaymentGatewaySettingsLayout

dmallory42 replied:
370a44a843 sorts out all these components. My original intention was to create a lightweight structure that can be used for any form page. In this commit, I took your advice about these components not being clear and reduced them to just one component with sub-sections, which I think is a lot clearer. I also added a comment explaining it.

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/gateway-settings-form/gateway-settings-form.tsx:
```diff
@@ -0,0 +1,18 @@
+	children: React.ReactNode;
+}
+
+export const GatewaySettingsForm: React.FC< GatewaySettingsFormProps > = ( {
```
```suggestion
export const PaymentGatewaySettingsForm: React.FC< GatewaySettingsFormProps > = ( {
```

dmallory42 replied:
Resolved by 370a44a843242f748568e6b9369d4d13cad5559b

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/field-placeholder/field-placeholder.tsx:
```diff
@@ -0,0 +1,27 @@
+
+export type FieldPlaceholderSize = 'small' | 'medium' | 'large';
+
+export const FieldPlaceholder = ( {
```
```suggestion
export const PaymentGatewayFieldPlaceholder = ( {
```

dmallory42 replied:
This can be used generically for any field, so IMO `FieldPlaceholder` is an appropriate name for it.

oaratovskyi replied:
OK, agree 🙂

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-section/payment-settings-section.scss:
```diff
@@ -0,0 +1,60 @@
+		max-width: 328px;
+
+		@include breakpoint( ">782px" ) {
+			flex: 0 0 33%;
```
please add comment about this behaviour

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-section/payment-settings-section.scss:
```diff
@@ -0,0 +1,60 @@
+
+		p {
+			font-size: 13px;
+			line-height: 17.89px;
```
unusual number too! Is that so in design?

dmallory42 replied:
I think this one was AI-generated weirdness. It should be a 20px line height. Fixed now!

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/components/payment-settings-section/payment-settings-section.tsx:
```diff
@@ -0,0 +1,28 @@
+ */
+import './payment-settings-section.scss';
+
+export const PaymentSettingsSection = ( {
```
```suggestion
export const PaymentGatewaySettingsSection = ( {
```

dmallory42 replied:
Removed in 370a44a843242f748568e6b9369d4d13cad5559b

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/offline/settings-payments-bacs.tsx:
```diff
@@ -0,0 +1,251 @@
+import { GatewaySettingsForm } from '~/settings-payments/components/gateway-settings-form';
+import { BankAccount } from '~/settings-payments/components/bank-accounts-list/types';
+
+export const SettingsPaymentsBacs = () => {
```
add please a comment about the component

oaratovskyi replied:
maybe call it direct bank transfer so it's more readable?

dmallory42 replied:
Updated in d0d09a984c. I think I prefer `bacs` because it is the same identifier used elsewhere in the code (more searchable), but I added a comment referring to it as direct bank transfer as well.

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/offline/settings-payments-bacs.tsx:
```diff
@@ -0,0 +1,251 @@
+							defaultCountry={ storeCountryCode }
+						/>
+					) }
+				</PaymentSettingsSection>
```
```suggestion
				</PaymentGatewaySettingsSection>
```
I suggest changing this and according styling - so it's more specific, otherwise folks can this it's related to the payment settings page

dmallory42 replied:
Removed in 370a44a843242f748568e6b9369d4d13cad5559b

oaratovskyi commented on plugins/woocommerce/client/admin/client/settings-payments/offline/settings-payments-cod.tsx:
```diff
@@ -0,0 +1,250 @@
+								'woocommerce'
+							) }
+							options={
+								codSettings.settings.enable_for_methods?.options
```
this does not work locally 🐛  idea: maybe move it into useEffect and use ready data here?

oaratovskyi reviewed (changes requested):
Thanks for this big work! It's in a good shape, but I believe we should fix it to make it work before merging
___
#### Review checklist
| Parameter  | Result |
| ------------- | ------------- |
| Approach  | ✅ All good |
| Implementation  | 👍  Almost good (left comments) |
| Automated tests  | ⚠️ Would be great to add some test to client components and newly introduced backend function |
| Naming  | 👍  Almost good (left comments) |
| Code comments  | 👍  Almost good (left a comment) |
| PR description  | ✅ Good |
| Testing instructions  | ✅ Good |
| Manual testing  | ⚠️ Faced some challenges during testing, see comment above | 

#### Scenarios tested
| Scenario  | Result | 
| ------------- | ------------- |
| Cash on delivery gateway  | ⚠️ Does not fully work. Need to address the shipping methods part |
| Check payments gateway  | ✅ Functionality worked correctly | 
| Direct bank transfer gateway  | ⚠️ Does not fully work. Need to address bank account adding part |

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Update getSortcodeLabel to use switch statement.

dmallory42 committed:
Lots of styling fixes, implementing new Settings component to wrap the content of a settings page.

dmallory42 committed:
Adding more documentation.

dmallory42 committed:
Styling and validation fixes.

dmallory42 committed:
Fixing bug with bank account settings and adding save state loader.

dmallory42 committed:
Fix onclose behaviour of ellipsismenu.

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Fix issue with loading of page if no shipping methods selected.

dmallory42 committed:
Fix lint errors.

dmallory42 committed:
Fix lint errors.

dmallory42 committed:
Lint fixes.

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Adding tests for new component.

dmallory42 committed:
Adding tests for the bank account modal and list.

oaratovskyi reviewed (commented):
Thanks, Daniel! It's in much better shape now, almost good!

Some items to sort:
- styling in cash on delivery shipping methods 
![image](https://github.com/user-attachments/assets/23d09f07-86c3-43ec-a382-ee32a220e6ca)
- Wrong check payments header title
![image](https://github.com/user-attachments/assets/00aeb603-af77-4b6c-bedb-3c2d38591b88)
- direct bank transfer Country is not saved

https://github.com/user-attachments/assets/0de68e97-3c02-4f6d-869a-766ad6770fdf

- "dirty" reloader when changed some data, saved and try to reload page with Ctrl +R

https://github.com/user-attachments/assets/6e8dafc2-d9b8-438e-a2c4-62ef88348364

dmallory42 committed:
Implement saving country for bank accounts.

dmallory42 committed:
Fix page title.

dmallory42 commented:
Thanks for the further reviews, @oaratovskyi - regarding your comments:

> styling in cash on delivery shipping methods

I think that sadly this won't be possible to address in this PR. We were able to identify a component, `TreeSelectControl` which provides the same functionality as the design, but the appearance is slightly different - while we can update it in future, I think we should do this as a follow-up issue since we'll possibly need to create a new component for it.

> Wrong check payments header title

Fixed. Thanks! 

> direct bank transfer Country is not saved

Great catch! Fixed - note that this relies on saving a new option value, so it will only work for newly added/saved bank accounts (legacy bank accounts will default to the store country until updated).

> dirty" reloader when changed some data, saved and try to reload page with Ctrl +R

I tried to fix this, but it's complicated because it's caused by the WP `mainform` which wraps the whole page. I think we can fix it, but I'd prefer to handle it in a new issue.

oaratovskyi reviewed (approved):
OK, thanks for addressing those! 
It's in a very good shape now - approving 🟢 
Please create small issues to address the follow ups with the styling and dirty data to address it next week 🙏

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Merge remote-tracking branch 'origin/trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Remove the navbar on offline PM pages.

dmallory42 commented:
I just pushed a small commit which hides the navbar on the offline methods pages as mentioned in the design feedback. 🙂

dmallory42 committed:
Adding logic to add border-bottom on last instance of class and fixing double borders.

dmallory42 committed:
Ensure visual alignment on settings page.

dmallory42 committed:
Minor copy updates.

dmallory42 committed:
Styling fixes.

dmallory42 committed:
Merge branch 'trunk' into dev/offline-method-updates-take-2

dmallory42 committed:
Add logic to detect changes, and fixed bug with enable_for_virtual field

dmallory42 commented:
Thanks for the great feedback @rogermattic - I have fixed most of it and I will leave comments on Figma for anything I couldn't address here so we can create issues for them.

dmallory42 committed:
Move order of bank fields.

dmallory42 committed:
Fix double border if no accounts.

dmallory42 committed:
Fix double border if no accounts.

dmallory42 committed:
Fix double border if no accounts.

dmallory42 committed:
Merge branch 'dev/offline-method-updates-take-2' of github.com:woocommerce/woocommerce into dev/offline-method-updates-take-2

dmallory42 committed:
Styling fix.

daledupreez commented:
@dmallory4, @woocommerce/moltres, and @rogermattic  we tripped over a subtle downstream problem stemming from these changes in QIT tests for the WooCommerce Stripe gateway. Prior to these changes, the original implementation created a `<label>` element for both the `title` and `label` attributes returned from the server, but we're now creating only a label for the `label` attribute. We tripped over this because the activation tests were using the label from the `title`.

I am not sure how major that is from a functional or usability perspective, as the act of clicking on the title section in the older settings screen is not something I'd normally expect to toggle a checkbox! This definitely feels like a "feature" of the old implementation, but I want us to make a conscious call as to what we will (and will not) support.

The pre-existing code that generates the label is here: https://github.com/woocommerce/woocommerce/blob/c46ae897aa0e2bb0f3290416967f60d7246b1e23/plugins/woocommerce/includes/abstracts/abstract-wc-settings-api.php#L461

alopezari commented:
Thanks for creating that draft fix @daledupreez!

Just wanted to ask for confirmation that there shouldn't be negative impact by accepting this change on how labels are generated, as it wasn't an actual requirement but a consequence of a recent implementation.

dmallory42 commented:
Thank you @daledupreez, that draft PR clarifies it for me! I think I was a little confused by the mention of the WooCommerce Stripe gateway as I was worried I'd inadvertantly broken something there! I understand now that the QIT tests were trying to access those elements by labels which have now changed.

> Just wanted to ask for confirmation that there shouldn't be negative impact by accepting this change on how labels are generated, as it wasn't an actual requirement but a consequence of a recent implementation.

> I think this might be a question for @dmallory42?

To confirm, the changes implemented in this PR are mainly a continuation of prior work to 'modernise' the payment settings pages, with an implementation that uses React components instead of the old method of injecting the page content via PHP. The changes in labelling are based on the design requirements which are referenced [here](https://github.com/woocommerce/woocommerce/issues/56364), but they shouldn't affect anything functionally - the settings page should allow a merchant to do exactly the same things as they could previously. 

So, in summary I think updating the tests to reflect the changes should be the only action we need to take here - and it shouldn't have a knock-on negative impact, because the way it was implemented means that only pages which are specifically 'marked' as having a React-powered version are generated in this way. Although, we should remember to check the QIT tests if we work on any pages in future.
